src/test/java/com/alphawallet/attestation/demo/VerifyAttestation.sol

```diff
@@ -0,0 +1,633 @@
+        uint256[2] memory check;
+
+        if (lhs[0] == 0 && lhs[1] == 0) { return false; } //early revert to avoid spending more gas
+
+        //Calculate RHS ECPoint
+        (check[0], check[1]) = extractXYFromPoint(pok.riddle);
+
+        //The hand optimised concat4 is more optimal than using abi.encodePacked (checked the gas costs of both)
+        bytes memory cArray = concat3Fixed(HPoint, pok.riddle, pok.tPoint);
+        uint256 c = mapTo256BitInteger(cArray);
+
+        //ECPoint riddle muliply by component hash
+        uint256[2] memory rhs = ecMul(c, check[0], check[1]);
+
+        if (rhs[0] == 0 && rhs[1] == 0) { return false; } //early revert to avoid spending more gas
+
+        //Add result of riddle.multiply(c) to point
+        (check[0], check[1]) = extractXYFromPoint(pok.tPoint);
+        rhs = ecAdd(rhs, check);
+
+        return ecEquals(lhs, rhs);
+    }
+
+    function ecMul(uint256 s, uint256 x, uint256 y) public view
+    returns (uint256[2] memory retP)
+    {
+        bool success;
+        // With a public key (x, y), this computes p = scalar * (x, y).
+        uint256[3] memory i = [x, y, s];
+
+        assembly
+        {
+        // call ecmul precompile
+        // inputs are: x, y, scalar
+            success := staticcall (not(0), 0x07, i, 0x60, retP, 0x40)
+        }
+
+        if (!success)
+        {
+            retP[0] = 0;
+            retP[1] = 0;
+        }
+    }
+
+    function ecInv(uint256[2] memory point) private pure
+    returns (uint256[2] memory invPoint)
+    {
+        invPoint[0] = point[0];
+        int256 n = int256(fieldSize) - int256(point[1]);
+        n = n % int256(fieldSize);
+        if (n < 0) { n += int256(fieldSize); }
+        invPoint[1] = uint256(n);
+    }
+
+    function ecAdd(uint256[2] memory p1, uint256[2] memory p2) public view
+    returns (uint256[2] memory retP)
+    {
+        bool success;
+        uint256[4] memory i = [p1[0], p1[1], p2[0], p2[1]];
+
+        assembly
+        {
+        // call ecadd precompile
+        // inputs are: x1, y1, x2, y2
+            success := staticcall (not(0), 0x06, i, 0x80, retP, 0x40)
+        }
+
+        if (!success)
+        {
+            retP[0] = 0;
+            retP[1] = 0;
+        }
+    }
+
+    function extractXYFromPoint(bytes memory data) public pure returns (uint256 x, uint256 y)
+    {
+        assembly
+        {
+            x := mload(add(data, 0x21)) //copy from 33rd byte because first 32 bytes are array length, then 1st byte of data is the 0x04;
+            y := mload(add(data, 0x41)) //65th byte as x value is 32 bytes.
+        }
+    }
+
+    function mapTo256BitInteger(bytes memory input) public pure returns(uint256 res)
+    {
+        bytes32 idHash = keccak256(input);
+        res = uint256(idHash);
+    }
+
+    function mapToCurveMultiplier(bytes memory input) public pure returns(uint256 res)
+    {
+        bytes memory nextInput = input;
+        do
```

Comment:
I think it is better to avoid the loop here to prevent gas leakage in case someone has not constructed the proof correctly s.t. c<curveOrder. Instead the smart contract should just terminate

Reply:
This function is no longer used. I left it in for reference. Since this file hasn't yet got a git history it's not lost by leaving it in the original commit. It can be removed later.

I'll update it though.